MimeRenderer Extensions

blink1073 opened the pull request:
Continuation of #2482.  Fixes #2354.

- [x] Add a declarative interface for mime renderer extensions
- [x] Add a registration method for mime renderer extensions
- [x] Add a widget and widget factory for mime renderer extensions
- [x] Use the registered mime renderers and add them to the rendermime and create
       widget factories if applicable
- [x] Thread the handling of mime extensions through to the build system
- [x] Rebase after #2480 and finish `MimeRenderer` implementation
- [x] Break out the `rendermime-interfaces` package
- [x] Add state restoration for mime renderer opened files
- [x] Finish tests

blink1073 commented:
TODO: break out docregistry handler into a separate plugin

ellisonbg commented:
I am testing this locally by updating Altair to use this MIME based renderer. So far working great. I will have some minor fixes to the vega package in the PR that I will submit as a PR to yours. I will then follow up with code review on the core code. Awesome to see it working though!

![screen shot 2017-06-20 at 8 41 59 pm](https://user-images.githubusercontent.com/27600/27366122-f26b1c6c-55f8-11e7-8235-d27adbc776a6.png)

ellisonbg commented on packages/docregistry/src/default.ts:
```diff
@@ -383,3 +399,213 @@ abstract class ABCWidgetFactory<T extends DocumentRegistry.IReadyWidget, U exten
+    let model = context.model;
+    let layout = this.layout as PanelLayout;
+    let data: JSONObject = {};
+    data[this._mimeType] = model.toString();
```
This may be the culprit in the Vega extension not working. I think it expects JSON data, rather than a string. We ran into this before, but I don't remember how we resolved it. We probably need to a figure out a way have a clear way of handling MIME types that have JSON or strings for data.

blink1073 commented:
The file data from the server is always text unless it is a notebook.  I added a `dataType` option that is threaded through to the mime renderer, and vega files are now rendering properly.
I also added state restoration for mime renderers.

ellisonbg commented:
Looks great now, merging!
